progressive_rollout_demo/lib/main.dart

```diff
@@ -0,0 +1,150 @@
+import 'package:shorebird_code_push/shorebird_code_push.dart';
+
+const groupKey = 'groupNumber';
+const firestoreCollectionName = 'rollouts';
+const firebaseDocumentName = 'rollouts';
+
+/// Assigns a random group number (1-100) to the user if one has not already
+/// been assigned.
+Future<void> assignGroupNumberIfNeeded() async {
+  final prefs = await SharedPreferences.getInstance();
+  if (!prefs.containsKey(groupKey)) {
+    final groupNumber = Random().nextInt(100) + 1;
+    await prefs.setInt(groupKey, groupNumber);
+  }
+}
+
+Future<void> main() async {
+  WidgetsFlutterBinding.ensureInitialized();
+
+  await Firebase.initializeApp();
+  await assignGroupNumberIfNeeded();
+
+  runApp(const MyApp());
+}
+
+class MyApp extends StatelessWidget {
+  const MyApp({super.key});
+
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+      theme: ThemeData(
+        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
+        useMaterial3: true,
+      ),
+      home: const MyHomePage(),
+    );
+  }
+}
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({super.key});
+
+  @override
+  State<MyHomePage> createState() => _MyHomePageState();
+}
+
+class _MyHomePageState extends State<MyHomePage> {
+  final _updater = ShorebirdUpdater();
+
+  int? _groupNumber;
+
+  /// Returns the current release version (e.g., 1.0.0+1).
+  Future<String> _releaseVersion() async {
+    final packageInfo = await PackageInfo.fromPlatform();
+    return '${packageInfo.version}+${packageInfo.buildNumber}';
+  }
+
+  /// Fetches the rollout percentage for the current release version
+  /// (e.g., 1.0.0+1) from Firestore.
+  ///
+  /// You can view the Firestore data in the console at:
+  /// https://console.firebase.google.com/u/0/project/progressive-rollout-demo/firestore/databases/-default-/data/~2Frollouts~2Frollouts
+  Future<int?> _fetchRolloutPercentage() async {
+    final collection = await FirebaseFirestore.instance
+        .collection(firestoreCollectionName)
+        .get();
+    final releaseVersion = await _releaseVersion();
+    return collection.docs.first.data()[releaseVersion] as int?;
+  }
+
+  /// Gets the group number (1-100) from shared preferences. We assume elsewhere
+  /// that this has been set before this widget is used.
+  Future<int?> _fetchGroupNumber() async {
+    final prefs = await SharedPreferences.getInstance();
+    return prefs.getInt(groupKey);
+  }
+
+  /// Determines the update track (beta or stable) based on the group number
+  /// and rollout percentage. If a user's group number is less than or equal to
+  /// the rollout percentage, they are assigned to the beta track. Otherwise,
+  /// they are assigned to the stable track. If there is no rollout percentage
+  /// for the current release version, all users are on the stable track.
+  Future<UpdateTrack> _updateTrack() async {
+    final (groupNumber, rolloutPercentage) = await (
+      _fetchGroupNumber(),
```

Comment:
Why are we always fetching the groupNumber even though it shouldn't change? It's a bit confusing that there are two `groupNumbers` and would be nicer to consolidate imo (e.g. load the group number and assign if needed once and use the cached value for all subsequent operations).

Reply:
I'd probably do a bunch of stuff differently if the point were to make this a production app (architecture, perf, etc.), but that's not the goal here—we're trying to make something easily readable and digestible.

Reply:
Since people will likely copy/paste this code the simplification/refactor I was proposing seems worthwhile. I was already confused by the fact that there are two `groupNumber`s during the PR review so I can imagine others would be tripped up by that as well.

Reply:
Done. A review of the doc would be helpful if you have a sec.